src/migrator/async.ts

```diff
@@ -0,0 +1,301 @@
+			initMigrationState.delayMS =
+				migration.delayMS || migratorEnv.asyncMigrationDefaultDelayMS;
+			initMigrationState.error_threshold =
+				migration.errorThreshold ||
+				migratorEnv.asyncMigrationDefaultErrorThreshold;
+		} else if (typeof migration === 'string') {
+			asyncRunnerMigratorFn = async (tx: Tx) =>
+				(await tx.executeSql(migration)).rowsAffected;
+		} else {
+			(sbvrUtils.api.migrations?.logger.error ?? console.error)(
+				`Invalid async migration object: ${JSON.stringify(migration, null, 2)}`,
+			);
+			continue;
+		}
+
+		await initMigrationStatus(setupTx, initMigrationState);
+
+		asyncMigrationSetup.push({
+			key,
+			initMigrationState,
+			asyncRunnerMigratorFn,
+		});
+	}
+
+	// Only after the setupTransaction has successfully finalized the asyncMigration runners will be
+	// created. When the transaction fails, the setup async migration entries in the DB will be
+	// rolled back automatically.
+	setupTx.on('end', () => {
+		for (const {
+			key,
+			initMigrationState,
+			asyncRunnerMigratorFn,
+		} of asyncMigrationSetup) {
+			const asyncRunner = async () => {
+				await sbvrUtils.db.transaction(async (tx) => {
+					await lockMigrations(tx, modelName, async () => {
+						const migrationState = await readMigrationStatus(tx, key);
+
+						if (!migrationState) {
+							// migration status is unclear stop the migrator
+							// or migration should stop
+							(sbvrUtils.api.migrations?.logger.info ?? console.info)(
+								`stopping async migration: ${key}`,
+							);
+							clearInterval(asyncScheduler);
+							return;
+						}
+						try {
+							// clear the interval to avoid re-triggering before finishing this migration query
+							clearInterval(asyncScheduler);
+
+							// sync on the last execution time between instances
+							// precondition: All running instances are running on the same time/block
+							// skip execution
+							if (migrationState.last_run_time) {
+								const durationSinceLastRun =
+									Date.now() - migrationState.last_run_time.getTime();
+								if (
+									(migrationState.is_backoff &&
+										durationSinceLastRun < initMigrationState.backoffDelayMS) ||
+									(!migrationState.is_backoff &&
+										durationSinceLastRun < initMigrationState.delayMS)
+								) {
+									// will still execute finally block where the migration lock is released.
+									return;
+								}
+							}
+							// set last run time and run counter only when backoff time sync between
+							// competing instances is in sync
+							migrationState.last_run_time = new Date();
+							migrationState.run_counter += 1;
+
+							let migratedRows = 0;
+							// here a separate transaction is needed as this migration may fail
+							// when it fails it would break the transaction for managing the migration status
+							await sbvrUtils.db.transaction(async (migrationTx) => {
+								migratedRows =
+									(await asyncRunnerMigratorFn?.(migrationTx)) || 0;
```

Comment:
This should presumably be
```suggestion
									(await asyncRunnerMigratorFn?.(migrationTx)) ?? 0;
```
?

Reply:
yes